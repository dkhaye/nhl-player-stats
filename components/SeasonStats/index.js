//
// components/SeasonStats/index.js
//
// Generate Player Statistics table given a stats array
//

import React from "react";

import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell, { tableCellClasses } from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import { styled } from '@mui/material/styles';

const StyledTableCell = styled(TableCell)(({ theme }) => ({
  [`&.${tableCellClasses.head}`]: {
    backgroundColor: theme.palette.common.black,
    color: theme.palette.common.white,
  },
  [`&.${tableCellClasses.body}`]: {
    fontSize: 14,
  },
}));

const StyledTableRow = styled(TableRow)(({ theme }) => ({
  '&:nth-of-type(odd)': {
    backgroundColor: theme.palette.action.hover,
  },
  // hide last border
  '&:last-child td, &:last-child th': {
    border: 0,
  },
}));

//
//  Function: SeasonStats
//  Description: renders the Table of the Player Statistics
//  Input:  stats - array of stat hashes containing year/team data
//  Output: Table of either Skater stats or Goalie stats (depending on existence of "wins") 
//
export default function SeasonStats({stats}) {
  if (stats.length > 0) {
    //Check if any year has "wins", treat as skater if not
    if (stats.findIndex((year) => {return (year.wins != null)}) == -1) {
      return(
        <div>
          <br/>
          <TableContainer component={Paper}>
            <Table aria-label="customized table">
              <TableHead>
                <TableRow>
                  <StyledTableCell>Season</StyledTableCell>
                  <StyledTableCell align="right">Team</StyledTableCell>
                  <StyledTableCell align="right">GP</StyledTableCell>
                  <StyledTableCell align="right">G</StyledTableCell>
                  <StyledTableCell align="right">A</StyledTableCell>
                  <StyledTableCell align="right">PIM</StyledTableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {stats.map((row) => (
                  <StyledTableRow key={row.year + row.team}>
                    <StyledTableCell component="th" scope="row">
                      {row.year}
                    </StyledTableCell>
                    <StyledTableCell align="right">{row.team}</StyledTableCell>
                    <StyledTableCell align="right">{row.games}</StyledTableCell>
                    <StyledTableCell align="right">{row.goals}</StyledTableCell>
                    <StyledTableCell align="right">{row.assists}</StyledTableCell>
                    <StyledTableCell align="right">{row.pim}</StyledTableCell>
                  </StyledTableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </div>
      );
    } else {
      //Player has "wins" and is a Goalie
      return(
        <div>
          <br/>
          <TableContainer component={Paper}>
            <Table aria-label="customized table">
              <TableHead>
                <TableRow>
                  <StyledTableCell>Season</StyledTableCell>
                  <StyledTableCell align="right">Team</StyledTableCell>
                  <StyledTableCell align="right">GP</StyledTableCell>
                  <StyledTableCell align="right">W</StyledTableCell>
                  <StyledTableCell align="right">L</StyledTableCell>
                  <StyledTableCell align="right">T/OT</StyledTableCell>
                  <StyledTableCell align="right">MIN</StyledTableCell>
                  <StyledTableCell align="right">GA</StyledTableCell>
                  <StyledTableCell align="right">SO</StyledTableCell>
                  <StyledTableCell align="right">GAA</StyledTableCell>
                  <StyledTableCell align="right">SV%</StyledTableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {stats.map((row) => (
                  <StyledTableRow key={row.year + row.team}>
                    <StyledTableCell component="th" scope="row">
                      {row.year}
                    </StyledTableCell>
                    <StyledTableCell align="right">{row.team}</StyledTableCell>
                    <StyledTableCell align="right">{row.games}</StyledTableCell>
                    <StyledTableCell align="right">{row.wins}</StyledTableCell>
                    <StyledTableCell align="right">{row.losses}</StyledTableCell>
                    <StyledTableCell align="right">
                      {row.ties ? row.ties : (row.ot ? row.ot : "-")}
                    </StyledTableCell>
                    <StyledTableCell align="right">
                      {row.timeOnIce.split(":")[0]}
                    </StyledTableCell>
                    <StyledTableCell align="right">
                      {row.goalsAgainst ? row.goalsAgainst : "-"}
                    </StyledTableCell>
                    <StyledTableCell align="right">
                      {row.shotsAgainst ? row.goalsAgainst : "-"}
                    </StyledTableCell>
                    <StyledTableCell align="right">
                      {row.goalAgainstAverage ? row.goalAgainstAverage.toFixed(2) : "-"}
                    </StyledTableCell>
                    <StyledTableCell align="right">
                      {row.savePercentage ? (row.savePercentage * 100).toFixed(2)+"%" : "-"}
                    </StyledTableCell>
                  </StyledTableRow>
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </div>
      );
    }
  } else {
    return(
      <div />
    );
  }
}
