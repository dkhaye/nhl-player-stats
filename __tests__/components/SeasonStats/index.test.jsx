/**
 * @jest-environment jsdom
 */

// __tests__/index.test.jsx

import React from 'react'
import { render, screen } from '@testing-library/react'
import SeasonStats from '@/components/SeasonStats'

beforeEach(() => {
  fetch.resetMocks();
});


describe('SeasonStats', () => {
  describe('When no stats', () => {
    const getParams = () => ({
      stats: []
    })

    it('renders empty', () => {
      render(<SeasonStats {...getParams()} />)

      const table = screen.queryByLabelText("customized-table");

      expect(table).toBeNull();
    })
  });

  describe('When stats with wins but missing stats', () => {
    const getParams = () => ({
      "stats": [{
        "year": null,
        "team": null,
        "timeOnIce" : null,
        "ot" : null,
        "shutouts" : null,
        "wins" : null,
        "losses" : null,
        "saves" : null,
        "powerPlaySaves" : null,
        "shortHandedSaves" : null,
        "evenSaves" : null,
        "shortHandedShots" : null,
        "evenShots" : null,
        "powerPlayShots" : null,
        "savePercentage" : null,
        "goalAgainstAverage" : null,
        "games" : null,
        "gamesStarted" : null,
        "shotsAgainst" : null,
        "goalsAgainst" : null,
        "powerPlaySavePercentage" : null,
        "evenStrengthSavePercentage" : null
      }]
    })

    it('renders goalie table', () => {
      render(<SeasonStats {...getParams()} />)

      const table = screen.getByRole("table");

      expect(table).toMatchSnapshot();
    })
  });


  describe('When stats with wins', () => {
    const getParams = () => ({
      "stats": [{
        "year": "2005 - 2006",
        "team": "Chicago Blackhawks",
        "timeOnIce" : "86:27",
        "ot" : 1,
        "shutouts" : 0,
        "wins" : 0,
        "losses" : 0,
        "saves" : 36,
        "powerPlaySaves" : 10,
        "shortHandedSaves" : 0,
        "evenSaves" : 26,
        "shortHandedShots" : 0,
        "evenShots" : 29,
        "powerPlayShots" : 12,
        "savePercentage" : 0.878049,
        "goalAgainstAverage" : 3.470214,
        "games" : 2,
        "gamesStarted" : 1,
        "shotsAgainst" : 41,
        "goalsAgainst" : 5,
        "powerPlaySavePercentage" : 83.33333333333334,
        "evenStrengthSavePercentage" : 89.65517241379311
      }]
    })

    it('renders goalie table', () => {
      render(<SeasonStats {...getParams()} />)

      const table = screen.getByRole("table");

      expect(table).toMatchSnapshot();
    })
  });

  describe('When stats without wins', () => {
    const getParams = () => ({
      "stats": [{
        "year": "2020 - 2021",
        "team": "Chicago Blackhawks",
        "timeOnIce" : "414:23",
        "assists" : 14,
        "goals" : 7,
        "pim" : 0,
        "shots" : 79,
        "games" : 19,
        "hits" : 3,
        "powerPlayGoals" : 2,
        "powerPlayPoints" : 6,
        "powerPlayTimeOnIce" : "75:33",
        "evenTimeOnIce" : "338:10",
        "penaltyMinutes" : "0",
        "faceOffPct" : 66.67,
        "shotPct" : 8.9,
        "gameWinningGoals" : 2,
        "overTimeGoals" : 0,
        "shortHandedGoals" : 0,
        "shortHandedPoints" : 0,
        "shortHandedTimeOnIce" : "00:40",
        "blocked" : 2,
        "plusMinus" : -5,
        "points" : 21,
        "shifts" : 426
      }]
    })

    it('renders skater table', () => {
      render(<SeasonStats {...getParams()} />)

      const table = screen.getByRole("table");

      expect(table).toMatchSnapshot();
    })
  });
})
