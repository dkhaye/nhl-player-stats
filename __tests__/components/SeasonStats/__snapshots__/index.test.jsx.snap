// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SeasonStats When stats with wins renders goalie table 1`] = `
<table
  aria-label="customized table"
  class="MuiTable-root css-rqglhn-MuiTable-root"
>
  <thead
    class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1wvnaxz-MuiTableCell-root"
        scope="col"
      >
        Season
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        Team
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        GP
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        W
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        L
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        T/OT
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        MIN
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        GA
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        SO
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        GAA
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        SV%
      </th>
    </tr>
  </thead>
  <tbody
    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
  >
    <tr
      class="MuiTableRow-root css-1enotpu-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
        scope="row"
      >
        2005 - 2006
      </th>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        Chicago Blackhawks
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        2
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        0
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        0
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        1
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        86
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        5
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        5
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        3.47
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        87.80%
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`SeasonStats When stats without wins renders skater table 1`] = `
<table
  aria-label="customized table"
  class="MuiTable-root css-rqglhn-MuiTable-root"
>
  <thead
    class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
  >
    <tr
      class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1wvnaxz-MuiTableCell-root"
        scope="col"
      >
        Season
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        Team
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        GP
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        G
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        A
      </th>
      <th
        class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1nslmh4-MuiTableCell-root"
        scope="col"
      >
        PIM
      </th>
    </tr>
  </thead>
  <tbody
    class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
  >
    <tr
      class="MuiTableRow-root css-1enotpu-MuiTableRow-root"
    >
      <th
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-132a6sb-MuiTableCell-root"
        scope="row"
      >
        2020 - 2021
      </th>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        Chicago Blackhawks
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        19
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        7
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        14
      </td>
      <td
        class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-1bhpkcd-MuiTableCell-root"
      >
        0
      </td>
    </tr>
  </tbody>
</table>
`;
